### PR TITLE
Bluetooth:  Fix type use of own_addr_type

### PR DESCRIPTION
**Pull request by Thalley**
The own_addr_type used for various HCI commands sometimes had a BT_HCI_OWN_ADDR_* type value or a BT_ADDR_* type value. Those are 2 different value spaces, and if the public address types would ever change, it would start have incorrect behavior.

Modified all assignments of own_addr_type to use the HCI defined values, and added a conversion function similar to the other Zephyr->HCI type functions we have. The function can today simply return the same value with some build asserts to avoid actually requiring any conversions.

### CODE REVIEW

**Comment by Thalley** [comment]
> It's not a pretty abstraction at all. I think it's not deserving a function, because that misleads the reader to think it's possible to simply convert an address type to an own address type value.

That is however the current behavior, so the reader is already mislead. I don't disagree that it's suboptimal, but it's not worse than what we have. 

> I suggest inlining the uses of this function again. Or is there some simple way to explain what this function does that I am just missing? (It's late in the day here. :P)

Hmm, this reads as you prefer implicitly incorrect (as per your own comment) conversion, rather than explicitly incorrect conversion. I do not see why the non-function conversion is better. 

The fact is that the current use of own_addr_type in the host implicitly assumes a whole lot of things. This PR removes the assumption and replaces with something that ensures some things (such as not setting the own address type to invalid values implicitly), and makes the implicit conversions of `*own_addr_type = addr->type` explicit. It also ensures that any changes to non-HCI address types from our API does not implicitly and silently change the behavior (e.g. if `BT_ADDR_LE_PUBLIC` is changed from `0` to `1` it would trigger an ASSERT rather than just sending `BT_HCI_OWN_ADDR_RANDOM` to the controller. 

I do not want to revert this is silently and implicitly converting values and then let Zephyr run on hopes and prayers. I am, however, more than plenty willing to improve this PR if you can think of ways to do this, but we do have to consider that the HCI spec uses 2 set of address type values that are not 100% interchangeable (as you've also pointed out). 

So in short, this PR does not change any current behavior, but it does prevent some (even more) wrong use of the conversions we already do, and fixes a lot of incorrect uses of the wrong address types in both the host and controller.

File: include/zephyr/bluetooth/hci_types.h [hunk 1]
```diff
@@ -34,6 +34,8 @@ extern "C" {
 #define BT_HCI_H4_ISO                   0x05    /* HCI ISO Data packet */
 
 /* Special own address types for LL privacy (used in adv & scan parameters) */
```
Comment (cvinayak) [inline include/zephyr/bluetooth/hci_types.h:36]: Change/Move comment as the newly added defines are not special, only the later are special due to the reason of being using only in adv & scan parameters.

File: subsys/bluetooth/controller/ll_sw/ull_adv.c [hunk 2]
```diff
@@ -455,8 +455,8 @@ uint8_t ll_adv_params_set(uint16_t interval, uint8_t adv_type,
 
 #if defined(CONFIG_BT_CTLR_PRIVACY)
 	adv->own_addr_type = own_addr_type;
-	if (adv->own_addr_type == BT_ADDR_LE_PUBLIC_ID ||
-	    adv->own_addr_type == BT_ADDR_LE_RANDOM_ID) {
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_adv.c:459]: Are the old values used for other purposes? if not, delete their definitions?

**Review by cvinayak — CHANGES_REQUESTED** [review]

**Comment by alwa-nordic** [comment]
> I am, however, more than plenty willing to improve this PR if you can think of ways to do this, but we do have to consider that the HCI spec uses 2 set of address type values that are not 100% interchangeable (as you've also pointed out).

I agree. Let's do a bit more than just inlinining. 

Consider https://github.com/zephyrproject-rtos/zephyr/pull/79645/files#diff-bbc763ec42cf05d416d946b19c10ab8b2921fe9ebc69ed993d71e41818bb99f4R1759-R1766.

I think this is more semantically meaningful:

```c
/* Set up own_addr_type to select the address of . */

if (bt_dev.id_addr[BT_ID_DEFAULT] == BT_ADDR_LE_RANDOM) {
    /* In case of Extended advertising: Assuming LE Set Advertising
     * Set Random Address command has not been used on this advertiser
     * set since it was created.
     */

    /* Set content of random address "RA" register. */
    set_random_address(&addr->a);

    /* Set address source select to register "RA". */
    *own_addr_type = BT_HCI_OWN_ADDR_RANDOM;
}

if (bt_dev.id_addr[BT_ID_DEFAULT] == BT_ADDR_LE_PUBLIC) {
    /* Assume BT_ID_DEFAULT is the public address read out from the
     * controller, so it's already in public address "PA" register.
     */

    /* Set address source select to register "PA". */
    *own_addr_type = BT_HCI_OWN_ADDR_RANDOM;
}
```

I still think it's not sensible to have a function `get_hci_own_addr_type: le_addr_type -> own_addr_type` because it does not capture semantic "reason", which is that we just set some register and want to select it.

File: subsys/bluetooth/controller/ll_sw/ull_adv.c (hunk 2 above)
Comment (alwa-nordic) [reply subsys/bluetooth/controller/ll_sw/ull_adv.c:459]: Out of scope of this PR, but `BT_ADDR_LE_PUBLIC_ID` and `BT_ADDR_LE_RANDOM_ID` are not Bluetooth address types. I think they are shoehorned here from some HCI event definition, and if that's the case, they should go with the other HCI definitions.

**Comment by Thalley** [comment]
> > I am, however, more than plenty willing to improve this PR if you can think of ways to do this, but we do have to consider that the HCI spec uses 2 set of address type values that are not 100% interchangeable (as you've also pointed out).
> 
> I agree. Let's do a bit more than just inlinining.
> 
> Consider https://github.com/zephyrproject-rtos/zephyr/pull/79645/files#diff-bbc763ec42cf05d416d946b19c10ab8b2921fe9ebc69ed993d71e41818bb99f4R1759-R1766.
> 
> I think this is more semantically meaningful:
> 
> ```c
> /* Set up own_addr_type to select the address of BT_ID_DEFAULT. */
> 
> if (bt_dev.id_addr[BT_ID_DEFAULT] == BT_ADDR_LE_RANDOM) {
>     /* In case of Extended advertising: Assuming LE Set Advertising
>      * Set Random Address command has not been used on this Advertising
>      * Set since it was created. Otherwise, that register would be selected
>      * by `BT_HCI_OWN_ADDR_RANDOM`.
>      */
> 
>     /* Set content of random address "RA" register. */
>     set_random_address(bt_dev.id_addr[BT_ID_DEFAULT].a);
> 
>     /* Set address source select to register "RA". */
>     *own_addr_type = BT_HCI_OWN_ADDR_RANDOM;
> }
> 
> if (bt_dev.id_addr[BT_ID_DEFAULT] == BT_ADDR_LE_PUBLIC) {
>     /* Assume BT_ID_DEFAULT is the public address read out from the
>      * controller, so it's already in public address "PA" register.
>      */
> 
>     /* Set address source select to register "PA". */
>     *own_addr_type = BT_HCI_OWN_ADDR_PUBLIC;
> }
> ```
> 
> I still think it's not sensible to have a function `get_hci_own_addr_type: le_addr_type -> own_addr_type` because it does not capture semantic "reason", which is that we just set some register and want to select it.

@alwa-nordic Please let me know if I understand your suggestion correctly: We remove any direct assignments from `*own_addr_type = type` (or with the function from this PR), and always assign `BT_HCI_OWN_ADDR_*` values directly to the `*own_addr_type` variables? 

If that is correctly understood, then I'm on board with that, i.e. no implicit or explicit conversions, but direct value assignments

**Comment by alwa-nordic** [comment]
> @alwa-nordic Please let me know if I understand your suggestion correctly: We remove any direct assignments from `*own_addr_type = type` (or with the function from this PR), and always assign `BT_HCI_OWN_ADDR_*` values directly to the `*own_addr_type` variables?
> 
> If that is correctly understood, then I'm on board with that, i.e. no implicit or explicit conversions, but direct value assignments

Yes, absolutely. To me that's the best way for this make sense.

**Comment by Thalley** [comment]
Removed the function and implicit conversions. 

I'm not 100% clear on the behavior in the second commit, so please read that carefully to see if I did it incorrectly (assuming that CI passes).

File: subsys/bluetooth/host/id.c [hunk 3]
```diff
@@ -1819,20 +1824,26 @@ int bt_id_set_scan_own_addr(bool active_scan, uint8_t *own_addr_type)
...
+				if (err) {
+					return err;
+				}
+
+				*own_addr_type = BT_HCI_OWN_ADDR_RANDOM;
+			} else if (bt_dev.id_addr[BT_ID_DEFAULT].type == BT_ADDR_LE_PUBLIC) {
+				*own_addr_type = BT_HCI_OWN_ADDR_PUBLIC;
+			} else {
+				LOG_DBG("Not changing the address for type %u",
+					bt_dev.id_addr[BT_ID_DEFAULT].type);
```
Comment (alwa-nordic) [inline subsys/bluetooth/host/id.c:1845]: This else is dead code. There are only two types of addresses.

File: subsys/bluetooth/host/id.c [hunk 4]
```diff
@@ -1747,9 +1751,12 @@ int bt_id_set_create_conn_own_addr(bool use_filter, uint8_t *own_addr_type)
 			if (err) {
 				return err;
 			}
-		}
 
-		*own_addr_type = addr->type;
+			*own_addr_type = BT_HCI_OWN_ADDR_RANDOM;
+		} else {
+			/* Privacy is not enabled, so if address type is not random, it's public */
```
[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/host/id.c]: Indeed. 

Why do we even have 
```
#define BT_ADDR_LE_PUBLIC_ID    0x02
#define BT_ADDR_LE_RANDOM_ID    0x03
```
?

They seem completely unused except for `bt_addr_le_from_str`

File: subsys/bluetooth/host/id.c (hunk 3 above)
Comment (Thalley) [reply subsys/bluetooth/host/id.c]: Should we add asserts in these cases, or can we somehow otherwise ensure that `type` is always one of the 2 values?

**Review by alwa-nordic — DISMISSED** [review]

**Review by jhedberg — DISMISSED** [review]

File: subsys/bluetooth/host/id.c (hunk 3 above)
Comment (alwa-nordic) [reply subsys/bluetooth/host/id.c]: Asserts or a CHECKIF would be appropriate here. Asserts if we have checks at API entry points (since we should only make assertions about our own code), otherwise CHECKIF.

File: subsys/bluetooth/host/id.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (alwa-nordic) [reply subsys/bluetooth/host/id.c]: They are not used directly anywhere. I think we can remove them. `ADDR_RESOLVED_BITMASK` is used instead.

However, it looks like we are abusing types:

https://github.com/zephyrproject-rtos/zephyr/blob/03959a20f7834ffe50a51df2b2e45ef87d1b610b/include/zephyr/bluetooth/hci_types.h#L2963

The above is not a valid `bt_addr_le_t`, since it can have that extra bit set. It seems addresses are cleaned up (using `bt_addr_le_copy_resolved`) most of the time before being given to the application. But worryingly, it looks like some code paths lead to this stuff bubbling up to the application.

File: subsys/bluetooth/host/id.c (hunk 3 above)
Comment (Thalley) [reply subsys/bluetooth/host/id.c]: Yeah, we could/should add those checks in the public API function that eventually get here to this internal function.

**Comment by Thalley** [comment]
Rebased to fix merge conflict

**Review by alwa-nordic — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by kruithofa — APPROVED** [review]

**Review by cvinayak — APPROVED** [review]

File: subsys/bluetooth/host/id.c [hunk 5]
```diff
@@ -1802,8 +1811,6 @@ int bt_id_set_scan_own_addr(bool active_scan, uint8_t *own_addr_type)
 			return err;
 		}
 	} else {
-		*own_addr_type = bt_dev.id_addr[0].type;
```
Comment (sean-madigan) [inline subsys/bluetooth/host/id.c:1805]: What should happen if `CONFIG_BT_SCAN_WITH_IDENTITY` is not enabled but `!is_adv_using_rand_addr())` is false? Prior to this change we would still set `*own_addr_type` here. Now `*own_addr_type` will be left undefined in this case.

Comment (Thalley) [reply subsys/bluetooth/host/id.c:1805]: I think you right. 
With `CONFIG_BT_SCAN_WITH_IDENTITY=n` and `is_adv_using_rand_addr() == false`, then `*own_addr_type` is not assigned. 

I think the previous assignment, `*own_addr_type = bt_dev.id_addr[0].type;`, was incorrect, as `id_addr[0].type` and the `BT_HCI_OWN_ADDR*` were in 2 different value spaces. 

Since this may happen when `CONFIG_BT_SCAN_WITH_IDENTITY=n` and `CONFIG_BT_PRIVACY=n` should the resulting address type be `BT_HCI_OWN_ADDR_RANDOM`?

Comment (Thalley) [reply subsys/bluetooth/host/id.c:1805]: I'm not sure why the code (and the original code) is guarded by `!is_adv_using_rand_addr()`, but it's old code it seems. @jhedberg @alwa-nordic any idea/suggestions?

Comment (jhedberg) [reply subsys/bluetooth/host/id.c:1805]: I think it would be good to move this discussion out of a closed PR into a new issue, to give it the attention it deserves. 

As for the background to this code, I can't say, at least not without doing some closer investigation into the matter.

Comment (Thalley) [reply subsys/bluetooth/host/id.c:1805]: https://github.com/zephyrproject-rtos/zephyr/issues/85202
